doc/libtcp4.2
.\" man page for Munts Technologies Linux Simple I/O Library
.\"
.\" Copyright (C)2016-2017, Philip Munts, President, Munts AM Corp.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions are met:
.\"
.\" * Redistributions of source code must retain the above copyright notice,
.\"   this list of conditions and the following disclaimer.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
.\" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.TH libtcp4 2 "12 June 2017" "version 1.0" "Linux Simple I/O Library"
.SH NAME
.B libtcp4
\-\- Linux Simple I/O Library: IPv4 TCP Module
.SH SYNOPSIS
.nf
.B #include <libtcp4.h>

.BI "void TCP4_resolve(char *" name ", int32_t *" addr ", int32_t *" error ");"

.BI "void TCP4_ntoa(int32_t " addr ", char *" dst ", int32_t " dstsize ", int32_t *" error ");"

.BI "void TCP4_connect(int32_t " addr ", int32_t " port ", int32_t *" fd ", int32_t *" error ");"

.BI "void TCP4_accept(int32_t " addr ", int32_t " port ", int32_t *" fd ", int32_t *" error ");"

.BI "void TCP4_server(int32_t " addr ", int32_t " port ", int32_t *" fd ", int32_t *" error ");"

.BI "void TCP4_close(int32_t " fd ", int32_t *" error ");"

.BI "void TCP4_send(int32_t " fd ", void *" buf ", int32_t " bufsize ", int32_t *" count ", int32_t *" error ");"

.BI "void TCP4_receive(int32_t " fd ", void *" buf ", int32_t " bufsize ", int32_t *" count ", int32_t *" error ");"

.fi
Link with
.BR -lsimpleio .
.SH DESCRIPTION
.nh
All functions return either
.B 0
(upon success) or an
.B errno
value (upon failure) in
.IR *error .
.PP
.B TCP4_resolve()
attempts to resolve an IPv4 address string passed in
.IR *name
(containing a domain name, a local host name like
.B localhost
or a dotted decimal address like
.BR 1.2.3.4 ).
Upon success, the 32-bit IPv4 address will be returned in
.IR *addr
.PP
.B TCP4_ntoa()
converts an IPv4 address to a dotted decimal address string. The
address of the destination buffer is passed in
.IR *dst
and its size, which must be at least 16 bytes, is passed in
.IR dstsize .
Upon success, the dotted decimal address string will be
returned in
.IR *dst .
.PP
.B TCP4_connect()
attempts to connect to a IPv4 TCP server.  The 32-bit IPv4 address is passed in
.IR addr
and the 16-bit TCP port number is passed in
.IR port .
Upon successful connection, a stream file descriptor will be returned in
.IR *fd .
.PP
.B TCP4_accept()
waits for an incoming connection request from a IPv4 TCP client.  Either
.B INADDR_ANY
may be passed in
.IR addr
to bind to (i.e. listen on) all network interfaces, or the 32-bit
IPV4 address of a particular network interface may be passed to bind
to only that interface.  The 16-bit TCP port number is passed in
.IR port .
Upon successful connection, a stream file descriptor will be returned in
.IR *fd .
.PP
.B TCP4_server()
operates like
.B TCP4_accept()
except that upon successful connection, the original server process forks
to create a new and separate connection handler process.  The
server process continues to listen for more connection requests,
without returning from
.B TCP4_server(),
while in the new connection handler process,
.B TCP4_server()
does return, with a stream file descriptor for the new connection returned in
.IR *fd .
.PP
.B TCP4_close()
closes a previously opened IPv4 TCP stream.
The stream file descriptor is passed in
.IR fd .
.PP
.B TCP4_send()
sends data to a IPv4 TCP stream.
The stream file descriptor is passed in
.IR fd .
The transmit buffer address is passed in
.IR buf " and"
its size is passed in
.IR bufsize .
Upon success, the number of bytes actually sent will be returned in
.IR *count .
.PP
.B TCP4_receive()
receives data from a IPv4 TCP stream.
The stream file descriptor is passed in
.IR fd .
The receive buffer address is passed in
.IR buf " and"
its size is passed in
.IR bufsize .
Upon success, the number of bytes actually received will be returned in
.IR *count .
.SH SEE ALSO
.BR libsimpleio "(2), " libevent "(2), " libgpio "(2), " libhidraw "(2), " libi2c "(2),"
.br
.BR liblinux "(2), " liblinx "(2), " libpwm "(2), " libserial "(2), " libspi "(2),"
.br
.BR libstream "(2), " libwatchdog "(2)"
.SH AUTHOR
Philip Munts, President, Munts AM Corp dba Munts Technologies


pascal/libtcp4.pas
{ FreePascal bindings for libsimpleio (http://git.munts.com/libsimpleio)      }

{ Copyright (C)2016-2017, Philip Munts, President, Munts AM Corp.             }
{                                                                             }
{ Redistribution and use in source and binary forms, with or without          }
{ modification, are permitted provided that the following conditions are met: }
{                                                                             }
{ * Redistributions of source code must retain the above copyright notice,    }
{   this list of conditions and the following disclaimer.                     }
{                                                                             }
{ THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" }
{ AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE   }
{ IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE  }
{ ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE   }
{ LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR         }
{ CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF        }
{ SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS    }
{ INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN     }
{ CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)     }
{ ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE  }
{ POSSIBILITY OF SUCH DAMAGE.                                                 }

UNIT libTCP4;

INTERFACE

  CONST
    INADDR_ANY      = $00000000;
    INADDR_LOOPBACK = $7F000001;

  PROCEDURE Resolve
   (hostname  : PChar;
    VAR addr  : Integer;
    VAR error : Integer); CDECL; EXTERNAL NAME 'TCP4_resolve';

  PROCEDURE NtoA
   (addr      : Integer;
    dst       : PChar;
    dstsize   : Integer;
    VAR error : Integer); CDECL; EXTERNAL NAME 'TCP4_ntoa';

  PROCEDURE Connect
   (addr      : Integer;
    port      : Integer;
    VAR fd    : Integer;
    VAR error : Integer); CDECL; EXTERNAL NAME 'TCP4_connect';

  PROCEDURE Accept
   (addr      : Integer;
    port      : Integer;
    VAR fd    : Integer;
    VAR error : Integer); CDECL; EXTERNAL NAME 'TCP4_accept';

  PROCEDURE Server
   (addr      : Integer;
    port      : Integer;
    VAR fd    : Integer;
    VAR error : Integer); CDECL; EXTERNAL NAME 'TCP4_server';

  PROCEDURE Close
   (fd        : Integer;
    VAR error : Integer); CDECL; EXTERNAL NAME 'TCP4_close';

  PROCEDURE Send
   (fd        : Integer;
    buf       : Pointer;
    size      : Integer;
    VAR count : Integer;
    VAR error : Integer); CDECL; EXTERNAL NAME 'TCP4_send';

  PROCEDURE Receive
   (fd        : Integer;
    buf       : Pointer;
    size      : Integer;
    VAR count : Integer;
    VAR error : Integer); CDECL; EXTERNAL NAME 'TCP4_receive';

IMPLEMENTATION

  USES
    initc;

  {$linklib libsimpleio}
END.
